woocommerce/woocommerce
# [Experiment] Additional field extensible sanitisation and validation handling

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Introduces new hooks for extensions to sanitise and validate registered checkout fields.

- Validation now uses an action hook with the error object passed by reference. This negates the need to compare the error object before and after hooks to ensure the object did not change
- Add sanitisation hook for custom sanitisation rules on additional fields
- Validation hook for custom validation rules on additional fields - key and value only
- Location validation hook for custom validation rules on additional fields where there are dependencies on the values of other fields in that location
- Validation applied to my account area

Closes #44019 



### How to test the changes in this Pull Request:



Developers should use this snippet this add some custom fields. Gov ID and email have validation methods.

<details>

```
add_action(
	'woocommerce_blocks_loaded',
	function() {
		woocommerce_blocks_register_checkout_field(
			array(
				'id'                => 'plugin-namespace/alt-email',
				'label'             => 'Alternative Email Field',
				'location'          => 'contact',
				'type'              => 'text',
				'required'          => true,
				'sanitize_callback' => function( $field_value ) {
					return sanitize_email( $field_value );
				},
				'validate_callback' => function( $field_value ) {
					if ( ! is_email( $field_value ) ) {
						return new \WP_Error( 'invalid_alt_email', 'Please ensure your alternative email matches the correct format.' );
					}
				},
			),
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'                => 'plugin-namespace/gov-id',
				'label'             => 'Government ID',
				'location'          => 'address',
				'type'              => 'text',
				'required'          => true,
				'sanitize_callback' => function( $field_value ) {
					return str_replace( ' ', '', $field_value );
				},
			),
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/confirm-gov-id',
				'label'    => 'Confirm Government ID',
				'location' => 'address',
				'type'     => 'text',
				'required' => true,
			),
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/leave-on-porch',
				'label'    => __( 'Please leave my package on the porch if I\'m not home', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'checkbox',
				'required' => true,
			),
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/location-on-porch',
				'label'    => __( 'Describe where we should hide the parcel', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'text',
			)
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/leave-with-neighbor',
				'label'    => __( 'Which neighbor should we leave it with if unable to hide?', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'select',
				'options'  => array(
					array(
						'label' => 'Neighbor to the left',
						'value' => 'left',
					),
					array(
						'label' => 'Neighbor to the right',
						'value' => 'right',
					),
					array(
						'label' => 'Neighbor across the road',
						'value' => 'across',
					),
					array(
						'label' => 'Do not leave with a neighbor',
						'value' => 'none',
					),
				),
			)
		);

		add_action(
			'woocommerce_blocks_sanitize_additional_field',
			function( $value, $key ) {
				if ( 'plugin-namespace/confirm-gov-id' === $key ) {
					return str_replace( ' ', '', $value );
				}
				return $value;
			},
			10,
			2
		);
		add_action(
			'woocommerce_blocks_validate_additional_field',
			function ( \WP_Error $errors, $field_key, $field_value ) {
				if ( 'plugin-namespace/gov-id' === $field_key ) {
					$match = preg_match( '/[A-Z0-9]{5}/', $field_value );
					if ( 0 === $match || false === $match ) {
						$errors->add( 'invalid_gov_id', 'Please ensure your government ID matches the correct format.' );
					}
				}
			},
			10,
			3
		);
		add_action(
			'woocommerce_blocks_validate_location_address_fields',
			function ( \WP_Error $errors, $fields, $group ) {
				if ( $fields['plugin-namespace/gov-id'] !== $fields['plugin-namespace/confirm-gov-id'] ) {
					$errors->add( 'gov_id_mismatch', 'Please ensure your government ID matches the confirmation.' );
				}
			},
			10,
			3
		);
	}
);
```

</details>

1. Add items to cart and go to checkout.
2. Enter something in the custom "additional email" field that is not an email, and enter some letters in the Gov ID fields. Ensure both Gov ID address fields are different.
3. After typing in Gov ID the checkout will refresh. You'll see an error notice with 2 bullets (formatting is wrong, and fields do not match). Correct this and ensure the errors go away.
4. Place order and it will throw another error this time about email. Correct this and place the order again. It should go through.
5. Check the confirmation page values match what you submitted.
6. Go to the My Account page.
7. Under addresses, edit an address. Try changing the Gov ID fields and submit with errors. You'll see an error notice if it fails validation.
8. If you refresh the page before fixing the validation errors, correct values are restored.
9. If you fix the validation errors, the new values should save to the account.
10. Under "account details" change "Alternative Email" to something else. Save. See error message.
11. Correct the email field and save. Change should persist.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
